Fixed Snippets Drag&Drop and added ability for Android Designer to do D&D

DavidKarlas opened the pull request:
Dropping things on `ITextView` is handled by different `IDropHandler`s. `IDropHandler` is chosen based on `[DropFormat ("XXX")]` and priority... So now when ToolboxPad starts dragging session we always insert `MonoDevelopToolBox` so our `MonoDevelopToolBoxDropHandler` is picked to handle dropping. Which then executes `ConsumeItem` for currently selected item on `Document` that is parent of `textView` that was dropped on.
I also added `IDragDataToolboxNode` interface which allows others e.g. Android Designer to define their own `DropFormat` and `byte[]` data when dragging starts that they want to pass to their own `IDropHandler`.

DavidKarlas commented:
I can't see way how to use GTK API for handling dropping things on CocoaTextView which is NSView.
Could you explain how this could be done?

Therzok commented on main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Toolbox/ToolboxDropHandler.cs:
```diff
@@ -0,0 +1,86 @@
+	[Export (typeof (IDropHandlerProvider))]
+	[DropFormat (DropFormat)]
+	[Name (nameof(MonoDevelopToolBoxDropHandlerProvider))]
+	class MonoDevelopToolBoxDropHandlerProvider : IDropHandlerProvider
```
```suggestion
	sealed class MonoDevelopToolBoxDropHandlerProvider : IDropHandlerProvider
```

Therzok commented on main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Toolbox/ToolboxDropHandler.cs:
```diff
@@ -0,0 +1,86 @@
+		}
+	}
+
+	class MonoDevelopToolBoxDropHandler : IDropHandler
```
```suggestion
	sealed class MonoDevelopToolBoxDropHandler : IDropHandler
```

mhutch commented:
Can you document IDragDataToolboxNode? Do we have a concrete use case for it?

mhutch commented:
Also, I feel like the editor should be aware of the toolbox, rather than the inverse. And will this compile on !mac?

DavidKarlas commented:
> Can you document IDragDataToolboxNode? Do we have a concrete use case for it?

Added documentation. Here is concrete use in Android Designer https://github.com/xamarin/designer/commit/a77a6305cb8

> Also, I feel like the editor should be aware of the toolbox, rather than the inverse. And will this compile on !mac?

I moved `ToolboxDropHandler.cs` to TextEditor.Cocoa integration. Sorry for breaking !mac, tnx for noticing that in time. I really wish we had CI for this :(

netonjm commented:
> I can't see way how to use GTK API for handling dropping things on CocoaTextView which is NSView.
> Could you explain how this could be done?

Ahh, gotcha I see your point, yeah... I remember I had same problem trying to start drag and drop from the native toolbox, I needed to hack the Gtk drag mechanism using same trick like Xwt and worked https://github.com/mono/xwt/blob/master/Xwt.Gtk/Xwt.GtkBackend/WidgetBackend.cs#L1146

But yes, remember it was too hacky, and if we could move to a native way without break GTK it would be great 👍

DavidKarlas commented:
This is not breaking GTK. Why do you think it breaks it?

abock commented:
@monojenkins backport release-8.2
